Replace distance with height for pre images

hewison-chris opened the pull request:
This changes the database field `distance` to `height` in the
`validator` table as it now is the block height for the `pre-image`.
This makes it much easier to work with preimages.
The `active` flag is also removed as we can deduce if an enrollment is
active from the `enrollment_height`. We keep all the previous enrollment
cycle records so that we know which validators were active during
previous cycles. This is needed for verifying signatures that we receive
after a new cycle has started.

Geod24 commented on source/agora/consensus/validation/Enrollment.d:
```diff
@@ -85,15 +85,14 @@ public string isInvalidReason (in Enrollment enrollment,
-        auto dist = height - (enroll_state.enrolled_height + enroll_state.preimage.distance);
-        assert(dist < ushort.max);
-        temp.distance = cast(ushort) dist;
+        auto dist = height - enroll_state.preimage.distance;
+        assert(dist < ulong.max);
+        temp.distance = dist;
```
You can just remove this no ?

hewison-chris replied:
Yes will do.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1096,7 +1127,8 @@ public class ValidatingLedger : Ledger
         const genesis_offset =  clock.networkTime - this.params.GenesisTimestamp;
         data.time_offset = max(genesis_offset, this.last_block.header.time_offset + 1);
-        log.trace("Going to nominate current time offset [{}] or newer. Genesis timestamp is [{}]", data.time_offset, this.params.GenesisTimestamp);
+        log.trace("Going to nominate current time offset [{}] or newer at network time [{}]. Genesis timestamp is [{}]",
+            data.time_offset, clock.networkTime, this.params.GenesisTimestamp);
```
Unrelated change ?

hewison-chris replied:
Will extract

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -421,12 +421,7 @@ public class Validator : FullNode, API
     {
         return new Clock((out long time_offset)
             {
-                // not enrolled - no need to synchronize clocks
-                if (!this.enroll_man.isEnrolled(&this.utxo_set.peekUTXO))
-                    return false;
```
Why was this removed ?

hewison-chris replied:
Firstly because we need to know the block height to know if we are enrolled. That is not that easy before we have a ledger.
Secondly because I do not see any harm if we sync anyway.

Geod24 replied:
Own commit then

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -536,29 +531,29 @@ public class Validator : FullNode, API
     ***************************************************************************/
 
-    protected void checkAndEnroll (Height block_height) @safe
+    protected void checkAndEnroll (Height height) @safe
```
Please minimize mixing refactoring (which the rename is) and functional changes.

hewison-chris replied:
Will extract into separate commit.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -1673,12 +1669,7 @@ public class TestValidatorNode : Validator, TestAPI
-                // not enrolled - no need to synchronize clocks
-                if (!this.enroll_man.isEnrolled(&this.utxo_set.peekUTXO))
-                    return false;
-
```
Ditto why

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -37,18 +37,9 @@ import d2sqlite3.sqlite3;
 
 import std.typecons : Tuple;
 
-public enum EnrollmentStatus : int
-{
-    Expired = 0,
-    Active = 1,
-}
-
```
And there was much rejoicing

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -163,27 +152,6 @@ public class ValidatorSet
         return null;
     }
 
-    /***************************************************************************
-
-        Returns:
-            the number of active enrollments, or 0 if there was a database error
-
-    ***************************************************************************/
-
-    public size_t count () @trusted nothrow
-    {
-        try
-        {
-            return this.db.execute("SELECT count(*) FROM validator_set WHERE active = ?", EnrollmentStatus.Active).
-                oneValue!size_t;
-        }
-        catch (Exception ex)
-        {
-            log.error("Error while calling ValidatorSet.count(): {}", ex);
-            return 0;
-        }
-    }
-
```
Unused I guess ? It's own commit then.

hewison-chris replied:
It was used but had a duplicate. I will extract into separate commit.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -204,24 +172,19 @@ public class ValidatorSet
     {
         try
         {
             () @trusted {
-                this.db.execute("DELETE from validator_set WHERE key = ? AND active = ?",
-                    EnrollmentStatus.Expired, enroll_hash);
-                this.db.execute("UPDATE validator_set SET active = ? WHERE key = ?",
-                    EnrollmentStatus.Expired, enroll_hash);
+                this.db.execute("DELETE from validator_set WHERE key = ?", enroll_hash);
             }();
         }
         catch (Exception ex)
```
Do we really want this function ? Feels odd to me to have both a primary key that takes into account the enrolled_height and a function to remove entries that is used to essentially keep the count of entries at 1 ?

hewison-chris replied:
It was only to keep the table row size to what is really needed. We should never need to query enrollments older than the previous cycle. I suppose redundant data cleanup can be done periodically by a task rather than each cycle.

Geod24 replied:
Don't we need it for block rewards ?

hewison-chris replied:
What shall we do with a slashed validator? It was being removed from the validator set which is probably not really what we want. However we do not want to treat it as an active validator after it has been slashed. A slashed flag?

Geod24 replied:
Short term yes. Long term we could JOIN with the UTXO set table (once made readable). The table won't have the missing UTXO so they won't show up.

hewison-chris replied:
Ok I'll add a `bool` field named `slashed`

ferencdg replied:
I think, we rely on the key_to_index and index_to_key maps in the Enrollment manager for block reward, because those maps are never cleared.

hewison-chris replied:
> I think, we rely on the key_to_index and index_to_key maps in the Enrollment manager for block reward, because those maps are never cleared.

At heights after the validator is slashed we should not add it to those maps though. I think it will be safer to mark as slashed in the table and leave the records there for validating older block signatures in case of the edge cases where we still get one for a slashed validator from a previous block.

ferencdg replied:
> > I think, we rely on the key_to_index and index_to_key maps in the Enrollment manager for block reward, because those maps are never cleared.
> 
> At heights after the validator is slashed we should not add it to those maps though. 

I expected that when I was implementing block reward calculation, so it shouldn't cause problems.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -236,18 +199,20 @@ public class ValidatorSet
-                " WHERE key = ? AND active = ?", enroll_hash, EnrollmentStatus.Active);
+            auto results = this.db.execute("SELECT enrolled_height FROM validator_set " ~
+                "WHERE key = ? AND enrolled_height <= ? AND enrolled_height >= ?",
+                enroll_hash, height, minEnrollmentHeight(height));
```
```suggestion
                enroll_hash, height, this.minEnrollmentHeight(height));
```

hewison-chris replied:
Added this to all.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -273,13 +238,14 @@ public class ValidatorSet
-                enroll_hash, EnrollmentStatus.Active);
+                "validator_set WHERE key = ? AND enrolled_height < ? " ~
+                "AND enrolled_height >= ?)", enroll_hash, height,
+                minEnrollmentHeight(height));
```
```suggestion
                this.minEnrollmentHeight(height));
```

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -302,13 +268,13 @@ public class ValidatorSet
-                "validator_set WHERE public_key = ? AND active = ?)", pubkey,
-                EnrollmentStatus.Active);
+                "validator_set WHERE public_key = ? AND enrolled_height < ? " ~
+                "AND enrolled_height >= ?)", pubkey, height, minEnrollmentHeight(height));
```
```suggestion
                "AND enrolled_height >= ?)", pubkey, height, this.minEnrollmentHeight(height));
```

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -340,8 +306,8 @@ public class ValidatorSet
-                WHERE enrolled_height < ? AND active = ? ORDER BY key ASC",
-                    height.value, EnrollmentStatus.Active);
+                WHERE enrolled_height < ? AND enrolled_height >= ? ORDER BY key ASC",
+                height, minEnrollmentHeight(height));
```
```suggestion
                height, this.minEnrollmentHeight(height));
```

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -236,18 +199,20 @@ public class ValidatorSet
-            auto results = this.db.execute("SELECT enrolled_height FROM validator_set" ~
-                " WHERE key = ? AND active = ?", enroll_hash, EnrollmentStatus.Active);
+            auto results = this.db.execute("SELECT enrolled_height FROM validator_set " ~
+                "WHERE key = ? AND enrolled_height <= ? AND enrolled_height >= ?",
```
IDK if it's just me but I'd find the condition more readable if the condition was inverted, with `>=` then `<=`. Because you should give the lower bound before the upper one. But not a big deal either way.

hewison-chris replied:
Yes I think I agree with you. I think it was just because I added the second check later. I will swap them over.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -273,13 +238,14 @@ public class ValidatorSet
             auto results = this.db.execute("SELECT EXISTS(SELECT 1 FROM " ~
-                "validator_set WHERE key = ? AND active = ?)",
-                enroll_hash, EnrollmentStatus.Active);
+                "validator_set WHERE key = ? AND enrolled_height < ? " ~
+                "AND enrolled_height >= ?)", enroll_hash, height,
```
Same comment about lower / upper bond

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -358,21 +324,21 @@ public class ValidatorSet
-                EnrollmentStatus.Active);
+            auto results = this.db.execute("SELECT key FROM validator_set " ~
+                "WHERE enrolled_height <= ? AND enrolled_height >= ? " ~
+                "ORDER BY key ASC", height, minEnrollmentHeight(height));
```
```suggestion
                "ORDER BY key ASC", height, this.minEnrollmentHeight(height));
```

Geod24 replied:
Love that `ORDER BY` BTW

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -386,38 +352,30 @@ public class ValidatorSet
-                this.db.execute("UPDATE validator_set SET active = ? WHERE enrolled_height <= ? AND active = ?",
-                    EnrollmentStatus.Expired, block_height - this.params.ValidatorCycle, EnrollmentStatus.Active);
+                this.db.execute("DELETE from validator_set WHERE enrolled_height <= ?",
+                    minEnrollmentHeight(Height(height - this.params.ValidatorCycle)));
```
```suggestion
                    this.minEnrollmentHeight(Height(height - this.params.ValidatorCycle)));
```

As mentioned earlier, I don't see the point of deleting if we always use `enroll_height`. And if we delete, I don't see the point of keeping `enroll_height` as part of the PK.

hewison-chris replied:
Refactored now so that it keeps the old enrollments but uses height to determine if enrolled.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -820,6 +794,24 @@ public class ValidatorSet
+
+    ***************************************************************************/
+
+    public Height minEnrollmentHeight (in Height height)
```
```suggestion
    public Height minEnrollmentHeight (in Height height) const scope @safe pure nothrow @nogc
```

hewison-chris replied:
Updated

Geod24 commented on source/agora/test/SlashingMisbehavingValidator.d:
```diff
@@ -155,8 +155,8 @@ unittest
     assert(block2.header.missing_validators.length == 1);
-    assert(nodes[0].getValidatorCount() == 5,
-        format!"Invalid validator count, current: %s"(nodes[0].getValidatorCount()));
+    assert(nodes[0].getValidatorCount(block2.header.height) == 5,
+        format!"Invalid validator count, current: %s"(nodes[0].getValidatorCount(block2.header.height)));
```
This should be cached

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -104,7 +103,7 @@ public class EnrollmentManager
     /// Validator set managing validators' information such as Enrollment object
     /// enrolled height, and preimages.
-    private ValidatorSet validator_set;
+    public ValidatorSet validator_set;  // Made public to ease transition to raise this to ledger
```
```suggestion
    public ValidatorSet validator_set;  // FIXME: Made public to ease transition to raise this to ledger
```

Geod24 reviewed (commented):
Nice

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -293,7 +285,7 @@ public class EnrollmentManager
             enroll = the enrollment data to add
             pubkey = the public key of the enrollment
-            block_height = current block height
+            height = current block height
```
If possible, this change should be in another commit. But it's your option.

linked0 commented on source/agora/consensus/data/PreImageInfo.d:
```diff
@@ -78,7 +78,7 @@ unittest
         utxo: utxo,
         hash: hash,
-        distance: 42,
+        distance: Height(42),
```
Shouldn't we have a different name for that, like `height`?

hewison-chris replied:
I think we can rename it as `height` but will probably do as last commit when it is working.

hewison-chris replied:
I renamed to height.

linked0 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -98,16 +89,16 @@ public class ValidatorSet
             "cycle_length INTEGER, enrolled_height INTEGER, " ~
-            "distance INTEGER, preimage TEXT, nonce TEXT, active INTEGER, " ~
-            "PRIMARY KEY (key, active))");
+            "distance INTEGER, preimage TEXT, nonce TEXT,
```
Shouldn't we have a different name for `distance`, like `preimage_height`?

hewison-chris replied:
Also probably `height` but not yet.

hewison-chris replied:
Also renamed and updated the DB doc.

linked0 reviewed (commented):
Great!
I will review more later.

Geod24 commented on .gitignore:
```diff
@@ -41,3 +41,5 @@ docs/
 # Auto-generated by the test that checks vtable offsets
 source/scpp/extra/DVMChecks.cpp
+source/agora/cli/version/version_build
+source/scpp/cpp_build
```
https://github.com/bosagora/agora/pull/1966

hewison-chris replied:
Change will drop when rebased.

Geod24 commented on doc/Database.md:
```diff
@@ -22,16 +22,16 @@ Our oldest table, only contain binary serialized data, should be changed to a re
+| height          | INTEGER  | ulong            |                                      |                                                    |
+| preimage        | TEXT     | Hash             |                                      |                                                    |
+| nonce           | TEXT     | Point            |                                      | The `R` used to sign the `Enrollment`              |
+| slashed_height  | INTEGER  | Height           |                                      | Height at which a validator is slashed or null     |
```
Why do we need to keep them in the validator table ?

hewison-chris replied:
We may still need to validate signatures for blocks before they were slashed. Hence the `slashed_height` as we need to be able to count active validators at every height.

Geod24 commented on source/agora/consensus/Quorum.d:
```diff
@@ -465,7 +465,8 @@ private NodeStake[] buildStakesDescending (const ref PublicKey filter,
     {
         UTXO value;
-        assert(finder(utxo_key, value), "UTXO for validator not found!");
+        if (!finder(utxo_key, value))
+            assert(0, "UTXO for validator not found!");
```
Seems irrelevant and part of Drey's PR

hewison-chris replied:
Will drop out on rebase.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -340,6 +343,7 @@ public class EnrollmentManager
     public Enrollment[] getEnrollments (in Height height, scope UTXOFinder peekUTXO)
         @trusted nothrow
     {
+        this.enroll_pool.removeExpired(height + 1);
```
This bit (and the function you introduced) could probably be its own commit / PR if you want to reduce the diff.

hewison-chris replied:
Will try but it may not work that easily with previous implementation.

hewison-chris replied:
I extracted the function into it's own commit but this change would not work with the previous logic.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -335,23 +327,36 @@ public class ValidatorSet
-            auto results = this.db.execute("SELECT key " ~
-                "FROM validator WHERE active = ? ORDER BY key ASC",
-                EnrollmentStatus.Active);
+            // ===========================DEBUG
+            import std.format;
+            auto rc = db.execute("select " ~
+                "key, public_key, cycle_length, enrolled_height, " ~
+                "height, preimage, nonce, slashed_height " ~
+                "from validator " ~
+                "order by height"
+                ).cached;
+            log.trace("%s\n%(%s\n%)".format(__FUNCTION__, rc));
+            // ===========================DEBUG
```
Leftover

hewison-chris replied:
Will remove

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -805,14 +759,35 @@ public class ValidatorSet
 
         return utxos;
     }
+
+    /***************************************************************************
+
+        Calculate the lowest height for an enrollemnt for the given height
+
+        Params:
+            height = block height
+
+        Returns:
+            min height for active enrollment at height given
+
+    ***************************************************************************/
+
+    public Height minEnrollmentHeight (in Height height) const scope @safe
+        pure nothrow @nogc
+    {
+        return Height(height <= this.params.ValidatorCycle ? 0
+            : height - this.params.ValidatorCycle);
+    }
```
Could probably be its own commit / PR too, and nice.

hewison-chris replied:
Yes will extract into commit.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -805,14 +759,35 @@ public class ValidatorSet
+
+    /***************************************************************************
+
+        Calculate the lowest height for an enrollemnt for the given height
```
```suggestion
        Calculate the lowest height an Enrollment had to appear at for it to be a validator at `height`
```

Does that sound clearer to you ?

hewison-chris replied:
Used as you suggested in next comment

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -805,14 +759,35 @@ public class ValidatorSet
+
+        Calculate the lowest height for an enrollemnt for the given height
+
+        Params:
+            height = block height
+
+        Returns:
+            min height for active enrollment at height given
```
Since the function is so trivial, you can use:
```D
Returns:
    Calculate the lowest height an Enrollment had to appear at for it to be a validator at `height`
```

Directly (so `Returns: description` without `Params` or the extra title).

hewison-chris replied:
Done

Geod24 reviewed (commented):
Few things, looks like you are still working on this ?

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -264,16 +265,17 @@ public class EnrollmentManager
 
         Params:
             enroll_hash = key for an enrollment block height
+            height = block height
```
It should be the first parameter.

hewison-chris replied:
Swapped.

hewison-chris committed:
Use height for pre images not distance

This changes the database field `distance` to `height` in the
`validator` table as it now is the block height for the `pre-image`.
This makes it much easier to work with preimages.
The `active` flag is also removed as we can deduce if an enrollment is
active from the `enrollment_height`. We keep all the previous enrollment
cycle records so that we know which validators were active during
previous cycles. This is needed for verifying signatures that we recieve
after a new cycle has started.

hewison-chris commented:
Fixed the Ledger tests. Next problem is with expected quorums.

Geod24 commented:
Need to remove the first commit

hewison-chris commented:
The third `QuorumShuffle` quorum checks are currently skipped as `createEnrollmentData()` in `base.d` doesn't create valid enrollments for end of 20 -> 40 block cycle. Not sure if it is worth fixing if we are going to remove it soon.
Otherwise I think all tests should pass now. 🤞

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -146,6 +147,7 @@ public class EnrollmentManager
         this.params = params;
         this.key_pair = key_pair;
         this.cycle = PreImageCycle(key_pair.secret, params.ValidatorCycle);
+        this.cycle.populate(key_pair.secret, true);
```
Populate shouldn't be needed since `seek` does the equivalent

hewison-chris replied:
This is gone now.

hewison-chris replied:
I just pushed a few mins ago.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -666,7 +666,7 @@ public class TestAPIManager
         this.setTimeFor(height);
         clients_idxs.each!(idx =>
-            retryFor(clients[idx].getBlockHeight() == height, timeout,
+            retryFor(clients[idx].getBlockHeight() >= height, timeout,
```
Looks like a bugfix, but at the same time should be justified in its own right. As currently we usually do an action, then expect one and not more block to be created, this I'm not 100% sure we want this.

hewison-chris replied:
I will revert. If we see in logs that height is actually more than expected we can investigate why.

hewison-chris replied:
Reverted

hewison-chris commented:
BTW, this still needs some changes extracted into new commits.

linked0 commented on source/agora/node/Ledger.d:
```diff
@@ -928,7 +921,7 @@ public class Ledger
     public Hash getValidatorRandomSeed (in Height height) nothrow
```
How about removing this function in this PR. This function is not used anymore anyway.

hewison-chris replied:
I added a commit to remove this.

hewison-chris committed:
Remove unused function

hewison-chris committed:
Increase timeout in unit test

The `RestoreSlashingInfo` network unit test sometimes fails with a 2
second retry timeout. Most tests use 5 seconds so updating it to be
consistent.

hewison-chris committed:
Simplify getting index of this node

Use an existing function to get the index of a node when getting the
index of the current node in a `slashing` check.

hewison-chris commented:
> The third `QuorumShuffle` quorum checks are currently skipped as `createEnrollmentData()` in `base.d` doesn't create valid enrollments for end of 20 -> 40 block cycle. Not sure if it is worth fixing if we are going to remove it soon.
> Otherwise I think all tests should pass now. 🤞

This test is no longer skipped as it has been fixed to work.

hewison-chris commented:
Now green but I will break out some more changes into smaller commits.

ferencdg commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1032,11 +1032,14 @@ extern(D):
-
-                log.trace("Combined consensus data: {}", data.prettify);
-                values ~= data;
+                {
```
what is the reason for not asserting? If a node considers a value as a candidate at some point, then it should always be valid from that point.

hewison-chris replied:
The missing validators can change over time so it can become invalid even though it was at some point in time.

ferencdg replied:
This is more like we are trying to ignore an existing bug with nomination/balloting.
If we  allow validateValue to return false, while previously returned true during the ballot phase for the same candidate, then you will get an assert in the cpp code and agora will crash anyway:
https://github.com/bosagora/agora/issues/1981#issuecomment-823864764

I have a suggestion in mind to tackle this in the right way: 

1 during 'nomination phase' it is okay to change a vote from 'false' to 'true', but impossible to change it from 'true' to 'false'. 
2 during 'balloting phase' it is not okay to change our votes in any ways.

*During nomination* 
1. If a node receives a candidate with **more** missing validators than the node knows, then he returns 'kInvalidValue'
2. If a node receives a candidate with **less** missing validators **or** with equal number of missing validators, but with different missing validators than the node knows, then it returns 'kInvalidValue' AND it will also query the preimages for all nodes in the quorum from the missing preimage(so he can update his missing validator array, and eventually vote with 'kValidValue' for the same candidate)
3. Otherwise return with 'kValidValue'

*During Balloting*

1. If there is any discrepancy between the missing validators we know and the candidate in question, we answer with 'kMaybeValid'
2. If there is no discrepancy between the missing validators we know and the candidate in question, we answer with 'kValidValue'


the above is just part of many nomination/slashing bugs improvements, and at some point I will solve them(including the corner case you mentioned)
https://github.com/bosagora/agora/issues/2022

hewison-chris replied:
But until what you mention is implemented then is it not better for combineCandidates to be able to choose another valid `consensusData` rather than asserting. These corner cases are what cause the build to sometimes fail. Are you sure that it will still crash if we choose another candidate from the set we are combining?

ferencdg replied:
If it causes test failures, then it is better to remove the assert, but maybe we can also put a link to the issue ticket #2022, so at some point we will solve the real issue.

Chosing an other candidate from the set will not not crash, but I think, it can indefinitely delay the balloting phase.

hewison-chris replied:
Perhaps this will not trigger anymore since the wait for preimages was added to many tests where it sometimes caused this. I will drop the commit from this PR and let's see if it hits the assert.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -424,7 +425,7 @@ public class EnrollmentManager
 
     public static Enrollment makeEnrollment (
-        in Hash utxo, in KeyPair key, in Hash seed, uint cycle_length, ulong offset)
+        in Hash utxo, in KeyPair key, in Hash seed, uint cycle_length, ulong offset = 0)
```
Hum why ? I'd imagine you'd merge it with the  above ?

Geod24 replied:
Discussed IRL.
